docsk8s/setting-up/ls-k8s-persistent-storage.asciidoc

```diff
@@ -1,16 +1,267 @@
+
+A workaround is to set a very long period like 1 year to `TerminationGracePeriodSeconds` to make sure {ls} get enough time to drain all events.
+
+[[persistent-storage-dlq]]
+==== Dead letter queue (DLQ)
+
+To enable {logstash-ref}/dead-letter-queues.html[dead letter queue], specify options in `logstash.yml`. The default path of DLQ is `/usr/share/logstash/data/dead_letter_queue`.
+
+[source,yaml]
+--
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: logstash-config
+data:
+  logstash.yml: |
+    api.http.host: "0.0.0.0"
+    dead_letter_queue.enable: true <1>
+  pipelines.yml: |
+    - pipeline.id: main <2>
+      path.config: "/usr/share/logstash/pipeline/main.conf"
+    - pipeline.id: dlq <3>
+      path.config: "/usr/share/logstash/pipeline/dlq.conf"
+--
+
+<1> Enable DLQ for all pipelines that use {logstash-ref}/plugins-outputs-elasticsearch.html[elasticsearch output plugin]
+
+<2> The `main` pipeline generates fail events to DLQ. Checkout the pipeline definition in the next session.
+
+<3> The `dlq` pipeline consumes the events in DLQ, fixes the error and sends the events to {es} again. Checkout the pipeline definition in the next session.
+
+[source,yaml]
+--
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: logstash-pipeline
+data:
+  main.conf: | <1>
+    input {
+      exec {
+        command => "uptime"
+        interval => 5
+      }
+    }
+    output {
+      elasticsearch { 
+        hosts => ["https://hostname.cloud.es.io:9200"]
+        index => "uptime-%{+YYYY.MM.dd}"
+        user => 'elastic'
+        password => 'changeme'
+      }
+    }
+  dlq.conf: | <2>
+    input {
+      dead_letter_queue {
+        path => "/usr/share/logstash/data/dead_letter_queue"
+        commit_offsets => true
+        pipeline_id => "main"
+      }
+    }
+    filter {
+        # Do your fix here
+    }
+    output {
+      elasticsearch { 
+        hosts => ["https://hostname.cloud.es.io:9200"]
+        index => "dlq-%{+YYYY.MM.dd}"
+        user => 'elastic'
+        password => 'changeme'
+      }
+    }
+--
+
+<1> An example pipeline that insert value to {es}. In the test environment, to generate fail event manually, use {ref}/indices-close.html[_close] API to close the index.
+
+<2> The pipeline use {logstash-ref}/plugins-inputs-dead_letter_queue.html[dead_letter_queue input plugin] to consume DLQ events. Add filter plugins to fix error causing fail insertion, for example, mapping error.
 
 [[persistent-storage-plugins]]
 ==== Plugins that require local storage to track work done
 
-Description...
+Here is the list of plugins require persistent storage to track the state of work. In order to have them work properly, use `StatefulSet` with `VolumeClaimTemplates`, checkout <<persistent-storage-statefulset>>.
```

Comment:
I don't think this is exhaustive, so maybe easier to give a rule of how to identify, and a list of "popular" plugins

```suggestion
Logstash plugins that are stateful will typically have some kind of `path` that needs to be configured, such as `sincedb_path` or `last_run_metadata_path`

Here is the list of popular plugins that will require persistent storage, and the use of a `StatefulSet` with `VolumeClaimTemplates`, checkout <<persistent-storage-statefulset>>.
```